File: VB/WebSite/App_Code/MyCode.vb
#Region "Using"

Imports Microsoft.VisualBasic
Imports System.Collections
Imports System.Collections.Generic
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports DevExpress.Web.ASPxEditors
Imports DevExpress.Web.ASPxTreeList
#End Region

Public Class MyDataCellTemplate
	Implements ITemplate
	Private tree As ASPxTreeList
	Public Sub New(ByVal tree As ASPxTreeList)
		Me.tree = tree
	End Sub

	Private Sub InstantiateIn(ByVal container As Control) Implements ITemplate.InstantiateIn
		Dim cellContainer As TreeListDataCellTemplateContainer = CType(container, TreeListDataCellTemplateContainer)
		If MyProvider.EditList.Contains(cellContainer.NodeKey) Then
			Dim box As New ASPxTextBox()
			container.Controls.Add(box)
			box.EnableViewState = False
			box.Value = cellContainer.Value
			box.Width = Unit.Percentage(100)
			box.ClientInstanceName = "myTreeEditor_" & cellContainer.NodeKey
		Else
			container.Controls.Add(New LiteralControl(cellContainer.Text))
		End If
	End Sub
End Class

Public Class MyItem
	Private m_id, m_parentId As Integer
	Private m_text As String

	Public Sub New()
	End Sub
	Public Sub New(ByVal id As Integer, ByVal parentId As Integer, ByVal text As String)
		m_id = id
		m_parentId = parentId
		m_text = text
	End Sub

	Public Property Id() As Integer
		Get
			Return m_id
		End Get
		Set(ByVal value As Integer)
			m_id = value
		End Set
	End Property
	Public Property ParentId() As Integer
		Get
			Return m_parentId
		End Get
		Set(ByVal value As Integer)
			m_parentId = value
		End Set
	End Property
	Public Property Text() As String
		Get
			Return m_text
		End Get
		Set(ByVal value As String)
			m_text = value
		End Set
	End Property
End Class

Public NotInheritable Class MyProvider

	Private Sub New()
	End Sub
	Public Shared ReadOnly Property EditList() As ArrayList
		Get
			Const key As String = "Q182424_1"
			If HttpContext.Current.Session(key) Is Nothing Then
				HttpContext.Current.Session(key) = New ArrayList()
			End If
			Return CType(HttpContext.Current.Session(key), ArrayList)
		End Get
	End Property


	Private Shared ReadOnly Property Data() As List(Of MyItem)
		Get
			Const key As String = "Q182424_2"
			If HttpContext.Current.Session(key) Is Nothing Then
				HttpContext.Current.Session(key) = CreateData()
			End If
			Return CType(HttpContext.Current.Session(key), List(Of MyItem))
		End Get
	End Property
	Private Shared Function CreateData() As List(Of MyItem)
		Dim data As List(Of MyItem) = New List(Of MyItem)()
		data.Add(New MyItem(0, -1, "Sales and Marketing"))
		data.Add(New MyItem(1, 0, "Field Office: Canada"))
		data.Add(New MyItem(2, 0, "Field Office: East Coast"))
		data.Add(New MyItem(3, -1, "Engineering"))
		data.Add(New MyItem(4, 3, "Consumer Electronics Div."))
		data.Add(New MyItem(5, 3, "Software Products Div."))
		Return data
	End Function

	Public Shared Function [Select]() As IEnumerable
		Return Data
	End Function
	Public Shared Sub Update(ByVal id As Integer, ByVal text As String)
		For Each item As MyItem In Data
			If item.Id = id Then
				item.Text = text
				Return
			End If
		Next item
	End Sub
End Class

File: CS/WebSite/App_Code/MyCode.cs
#region Using
using System.Collections;
using System.Collections.Generic;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using DevExpress.Web.ASPxEditors;
using DevExpress.Web.ASPxTreeList;
#endregion

public class MyDataCellTemplate : ITemplate {
	ASPxTreeList tree;
	public MyDataCellTemplate(ASPxTreeList tree) {
		this.tree = tree;
	}

	void ITemplate.InstantiateIn(Control container) {
		TreeListDataCellTemplateContainer cellContainer = (TreeListDataCellTemplateContainer)container;
		if(MyProvider.EditList.Contains(cellContainer.NodeKey)) {		
			ASPxTextBox box = new ASPxTextBox();
			container.Controls.Add(box);
			box.EnableViewState = false;
			box.Value = cellContainer.Value;
			box.Width = Unit.Percentage(100);
			box.ClientInstanceName = "myTreeEditor_" + cellContainer.NodeKey;
		} else {
			container.Controls.Add(new LiteralControl(cellContainer.Text));
		}
	}
}

public class MyItem {
	int m_id, m_parentId;
	string m_text;

	public MyItem() {
	}
	public MyItem(int id, int parentId, string text) {
		m_id = id;
		m_parentId = parentId;
		m_text = text;
	}

	public int Id { get { return m_id; } set { m_id = value; } }
	public int ParentId { get { return m_parentId; } set { m_parentId = value; } }
	public string Text { get { return m_text; } set { m_text = value; } }
}

public static class MyProvider {

	public static ArrayList EditList {
		get {
			const string key = "Q182424_1";
			if(HttpContext.Current.Session[key] == null)
				HttpContext.Current.Session[key] = new ArrayList();
			return (ArrayList)HttpContext.Current.Session[key];
		}
	}


	static List<MyItem> Data {
		get {
			const string key = "Q182424_2";
			if(HttpContext.Current.Session[key] == null)
				HttpContext.Current.Session[key] = CreateData();
			return (List<MyItem>)HttpContext.Current.Session[key];
		}
	}
	static List<MyItem> CreateData() {
		List<MyItem> data = new List<MyItem>();
		data.Add(new MyItem(0, -1, "Sales and Marketing"));
		data.Add(new MyItem(1, 0, "Field Office: Canada"));
		data.Add(new MyItem(2, 0, "Field Office: East Coast"));
		data.Add(new MyItem(3, -1, "Engineering"));
		data.Add(new MyItem(4, 3, "Consumer Electronics Div."));
		data.Add(new MyItem(5, 3, "Software Products Div."));
		return data;
	}

	public static IEnumerable Select() {
		return Data;
	}
	public static void Update(int id, string text) {
		foreach(MyItem item in Data) {
			if(item.Id == id) {
				item.Text = text;
				return;
			}
		}
	}
}